### DISCUSSION TITLE
CW QSO directly from Cloudlog

### DISCUSSION THREAD

**Discussion by ok1cdj** [magicbug/Cloudlog#814] [Ideas / Feature Requests]
Hello all,
Cloudlog is my primary log  for daily use and it will be nice to make CW QSO directly from log.
I made network morse keyer based on ESP32 with simple web interface and usage is really simple:

http://XX.XX.XX.XX/get?message=CQ DE OK1CDJ OK1CDJ K&speed=25&API_KEY=XXXXXXXX

and keyer returns OK when message is accepted.

Is there someone who is also interested in this feature ? 
Can someone  help with integration in to Cloudlog ??

Have a nice day.

Ondra OK1CDJ

**Comment by magicbug** [comment]
Happy to help intergrate although I'll need a method to test it but I can see this being popular

**Comment by daccle** [comment]
Hi Ondra!

Cloud share the design of the morse keyer or provide at least a mockup server, so that one can test?

Best, Daniel DC2ZP

**Comment by ok1cdj** [comment]
Hello all, code is made for M5StickC now, but can be used on any ESP32 module (WIFI or Ethernet). I will put whole code to github as open source in next few days.
I can make simple mockup server for testing. I can make any changes in interface for easy integration. 
We can discuss what we need on Cloudlog side:

1.  Configuration - Name of keyer, IP address, port and API key   - Q: more keyers based on station profiles ??
2.  Text macro - some system with variables ??
3.  Buttons mapped to macros in Live QSO window
4.  Keyboard shortcuts - Q: configurable or not ??


73 Ondra OK1CDJ

**Comment by ok1cdj** [comment]
Code for keyer is available: https://github.com/ok1cdj/M5-ESP32-keyer

73 Ondra OK1CDJ

**Comment by magicbug** [comment]
I've ordered an M5stick from a UK seller (seems they are about £14) to try intergrate this well.

Think there needs to be a few things here

- hardware selection for Morse code, this would need to be user level the meaning later other CW keyers can be supported
- user defined macros for either general or contesting (@m0pcb would kill me not having these seperate) so this would need a UI
- buttons displayed on both interfaces.

Does that sound right?

**Reply by ok1cdj** [reply]
Sounds good. Maybe some hotkeys will be great.

Any sugestions to my Firmware are welcome.

73 Ondra OK1CDJ

**Comment by m0pcb** [comment]
Hotkeys, yes. I guess configurable?

Also would be nice for cloudlog to also support winkey interfaces as well
as that awesome looking keyer you built Ondra!

73, Iain AD5XI / M0PCB

**Reply by ok1cdj** [reply]
There is new WEB Serial API - https://codelabs.developers.google.com/codelabs/web-serial#0
With this i possible implement serial connected devices like WINKEY,
but for my opinion, network devices is better to use from cloud.

73 Ondra OK1CDJ

**Reply by magicbug** [reply]
That's exactly hope that webserial can be used, it's possible it will also allow better cat functions for those using location machines.

Without this, it needs similar to cat apps where I have in the middle apps which is ok.

What I need is the winkey protocol, asked the developer for the DLL over a year ago and got zero reply.

**Reply by ok1cdj** [reply]
Winkey protocol is not open, you can find C++ implementation in K3NG keyer 
https://github.com/k3ng/k3ng_cw_keyer
or Perl  https://github.com/N0NB/winkeydaemon
or Pascal https://github.com/ok2cqr/winkeyer_server

Have a nice evening.

73 Ondra OK1CDJ

**Comment by mbridak** [comment]
Just some quick notes on winkeyer use:

open serial to 1200 baud.
dtr true
rts false
wait one second.
send byte values x00 x02 to 'open' the device.
device will reply w/ version.
send device bytes x0e x44 to set mode (echo back sent cw)
send device bytes x02 x14 to set speed to 20

to send cw after setup just send uppercase byte encoded data to serial port.

to send prosign, send byte x1b then two byte encoded uppercase chars.

The device can send status bytes back at any time.
to decode status you'd logic and the status with xc0.

if statusbyte && xc0 == xc0: it's a status byte
if statusbyte && xc0 == x80: new speedpot setting
else it's an echoed character.

**Reply by magicbug** [reply]
Thanks, I've found a useful c# library too so I just need to find some time to implement it.

**Comment by ok1cdj** [comment]
Here is example how to control K3NG Keyer over serial CLI directly from browser using webserial API in javascript
https://github.com/ok1cdj/K3NG-keyer-serial-terminal

I think this can be very easily implemented to cloudlog.

73 Ondra OK1CDJ

**Reply by magicbug** [reply]
Oh looks fantastic, I'm guessing it would be possible todo keying of RTS or CTS for CW keying too with webserial not something I've tried!

**Reply by ok1cdj** [reply]
Yes, webserial API alow to control RTS and DTR signals, but there can be problem with precision of timing. Keyer with serial input as string can be done with 5 EUR arduino with K3NG firmware for example.

73 Ondra OK1CDJ
